src/katsdpmodels/fetch/requests.py

```diff
@@ -322,68 +279,19 @@ def get(self, url: str, model_class: Type[_M], *,
         requests.exception.RequestException
             Any exceptions raised by the underlying session.
         """
-        original_url = url
-        url = self.resolve(url)[-1]
-        if url in self._model_cache:
-            model = self._model_cache[url]
-            if model_class.model_type != model.model_type:
-                raise models.ModelTypeError.with_urls(
-                    f'Expected a model of type {model_class.model_type!r}, '
-                    f'not {model.model_type!r}',
-                    url=url, original_url=original_url)
-            if not isinstance(model, model_class):
-                raise TypeError('model_class should be the base class for the model type')
-            return model
-
-        try:
-            if lazy:
-                fh, url, content_type = self._get_lazy(url)
-                checksum: Optional[str] = None
-            else:
-                fh, url, content_type, checksum = self._get_eager(url)
-        except models.ModelError as exc:
-            exc.original_url = original_url
-            exc.url = url
-            raise
-        if content_type == 'application/octet-stream':
-            # This is a generic/fallback content type that doesn't convey any
-            # useful information. The load_file method should use the
-            # extension instead.
-            content_type = None
-
-        try:
-            with contextlib.ExitStack() as exit_stack:
-                exit_stack.callback(fh.close)
-                new_model = model_class.from_file(fh, url, content_type=content_type)
-                exit_stack.pop_all()   # new_model now owns fh, or has closed it
-        except models.ModelError as exc:
-            exc.original_url = original_url
-            exc.url = url
-            raise
-        except Exception as exc:
-            raise models.DataError.with_urls(
-                f'Failed to load model from {url}: {exc}',
-                url=url, original_url=original_url) from exc
-        new_model.checksum = checksum
-        self._model_cache[url] = new_model
-        return new_model
-
-    def __enter__(self: _F) -> _F:
-        return self
-
-    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
-        self.close()
+        return self._run(self._get(url, model_class), lazy=lazy)
 
 
 def fetch_model(url: str, model_class: Type[_M], *,
-                session: Optional[Session] = None) -> _M:
-    """Convenience function for retrieving a single model.
+                session: Optional[requests.Session] = None) -> _M:
```

Comment:
This can surely be shared between aiohttp and requests

Reply:
One is a coroutine, and calls a coroutine, while the other does not. Possibly one could do complicated things with introspection and futures, but it would almost certainly add more code than it would save and it would be harder to understand.